### ISSUE TITLE
[RFC] Cross-Process Performance Analysis: Straggler Detection

### ISSUE THREAD

**Issue by wayi1** [pytorch/pytorch#71303]
## Motivation: Limitation of Existing Profiling Approach
To conduct PyTorch distributed training performance analysis, currently a recommended way is profiling by using [PyTorch profiler](https://pytorch.org/docs/stable/profiler.html) or [Nsight](https://docs-dev.nersc.gov/cgpu/software/profiling/). As a result, every profiled process can output a trace file, and users usually only need to take one or two representative files to analyze (mostly for the case of data parallelism).

However, one limitation of such approach is no easy support for cross-process performance analysis. A prominent example is **straggler detection**, because it's hard to identify straggling by only looking into a small number of individual trace files, and if the training runs at a relatively large scale, it's also impractical for users to eyeball a large number of trace files.

## Impact
Note that as we scale out trainings to more and more machines nowadays, straggler issues can become increasingly common, especially given that the most well-adopted PyTorch distributed training implementation, DDP is highly sensitive to stragglers -- the allreduce-based gradient synchronization has to wait for the slowest process at every iteration.

I believe that straggler issues can occur in various cases as DDP training is scaled out, and detecting and mitigating stragglers can have a high impact for improving large-scale training performance. A few cases:

1. Embedding table lookup can be expensive with large embeddings, especially these embedding tables may need to be partitioned across machines and hence incur all-to-all communication overheads. Such workloads may have a large variance across processes in the forward pass and hence cause stragglers. This can be found in training large-scale recommendation systems. (This one can be quite relevant to Ads companies like Meta: @zhaojuanmao, @rohan-varma)
2. If a massive amount of training data is used for training a relatively small model, the I/O cost can become a bottleneck. In this case, stragglers can occur if the data size of the input examples/records have a very large variance. Reading the input data from (potentially multi-homed) cloud storage may make this even worse.
3. If a customized data loader has a very computationally expensive `transform` phase, and a high variance in the computation amount of on-the-fly data transformation may also contribute to straggling.

On the other hand, PyTorch users may not be even aware of such pattern ([example](https://github.com/pytorch/pytorch/issues/59433)).

## Detailed Requests
Can PyTorch provide an easy way for straggler detection? Particularly:

1. It should be easy to install such a detection component, and there should be no intrusive code change.
2. Straggling can be defined programmatically, and a metric can be reported to measure the extent of straggling.
3. Report whether the stragglers are the same across different iterations. If so, the straggling is more likely to be host/machine-specific, or it may imply the lack of sufficient data shuffling, etc.

With such feature, it will be much easier for industry powerful users to run fleet-wide straggler detections, and PyTorch users can be more alert to stragglers.

## Possible Implementations

1. We can develop a new type of [debugging DDP communication hook](https://pytorch.org/docs/stable/ddp_comm_hooks.html#debugging-communication-hooks). In the hook, before allreduce is kicked off, first measure the start time in the hook state, and then summarize different start times across processes at the same iteration. However, in the hook, it's impossible to also measure the total runtime of an iteration. To evaluate the extent of straggling, may need to bring some side inputs (e.g., regarding the total runtime of an iteration) when parsing a hook state that records allreduce starts times.
2. We can build an analysis layer on top of existing DDP logging, which has already collected all the raw data needed for this purpose. CC: @zhaojuanmao, @rohan-varma

**Comment by rohan-varma** [comment]
I wonder if we need to develop something entirely new here, or if we just need to take something like `_get_ddp_logging_data` and make the data/insights more relevant/easily digestible by end users. For example, straggler detection can be handled with `_get_ddp_logging_data` (admittedly in a not very user friendly manner) today.

Longer term, ideally outputs from something like `_get_ddp_logging_data` should be visible in Tensorboard for OSS users. There is already a distributed profiling view developed by Microsoft: https://pytorch.org/blog/pytorch-profiler-1.9-released/, we could possibly look into adding to this.

**Comment by wayi1** [comment]
> will the use case not be satisfied by using DDP APIs such as get_logging_data or enabling this detection with TORCH_DISTRIBUTED_DEBUG?

I think both routes could work, so I don't have any preference here. Personally I would not prefer the name of **TORCH_DISTRIBUTED_DEBUG** mode in this context, because this mode is mainly designed for debugging distributed training failures that normally lead to a certain uninformative error (e.g., NCCL timeout). I am not sure if we want to view straggler as a **bug** here. This may cause some user confusion.

> Longer term, ideally outputs from something like _get_ddp_logging_data should be visible in Tensorboard for OSS users.

This can be a good idea. Just check if we can have any more lightweight solution before reaching this stage. Additionally, would love to see if we can have some analytics on top of the raw data to summarize the extent of straggling. This can help a lot on a fleet-wide performance analysis.

> There is already a distributed profiling view developed by Microsoft: https://pytorch.org/blog/pytorch-profiler-1.9-released/, we could possibly look into adding to this.

This can be a good entry point. I have used this distributed view on TensorBoard, and I can confirm it does not have any support on the straggler detection yet.

**Comment by mrshenli** [comment]
I guess the approach might depend on the scope. If you only care about DDP stragglers, having a custom comm hook totally makes sense. If you would like to cover all collective-based features (DDP/ZeRO/FSDP), having a context manager might be more generic. It can be sth like below, where `sd.summary()` can tell how much time is wasted due to straggler on each process.

```python

with torch.distributed.straggler_detection() as sd:
    # some training iterations

if rank == 0:
    print(sd.summary())
```

**Comment by wayi1** [comment]
> I guess the approach might depend on the scope. If you only care about DDP stragglers, having a custom comm hook totally makes sense. If you would like to cover all collective-based features (DDP/ZeRO/FSDP), having a context manager might be more generic. It can be sth like below, where `sd.summary()` can tell how much time is wasted due to straggler on each process.
> 
> ```python
> with torch.distributed.straggler_detection() as sd:
>     # some training iterations
> 
> if rank == 0:
>     print(sd.summary())
> ```

Thanks for the suggestion @mrshenli! 

> I guess the approach might depend on the scope. If you only care about DDP stragglers, having a custom comm hook totally makes sense.

Currently my end will only need to target DDP scope.

> If you would like to cover all collective-based features (DDP/ZeRO/FSDP), having a context manager might be more generic.

Personally I probably will prefer an alternate approach, which can be a separate module to run some offline analysis over traces collected by torch profiler. A few reasons:

1. In practice, users may want to run profiling first instead of directly running straggler detection, so it would be nice if we can incorporate such a new context manager into profiler context manager. We still have the advantage of not being restricted to any specific collective communication.
2. Microsoft folks working on profiler told me that they plan to upstream a module PT1.11, which can provide an API to conveniently export perf stats from trace files. This makes such route more viable in the future.
3. This approach can not only support straggler detection, but also other trace-based analysis as well.

On the other hand, I still think a debug DDP comm hook for straggler detection can still make sense.

1. It's more lightweight than enabling profiling, which is often more than adding a context manager for multi-node training in practice. Profiling often requires other setup like uploading trace files to a cloud storage w/o exceeding the rate limit.
2. It's much easier for industrial users to deploy a DDP comm hook than a profiler context manager at the fleet wide, because setting up a DDP comm hook can be totally decoupled from the training loop, which means such setup could be done in one place. In contrast, for industrial users every use case usually has a different training loop in different files, the workload will be much larger, unless a higher-level API (probably based on Lightning or Ignite) has unified these loops in one place.

**Comment by rohan-varma** [comment]
Adding hi-pri as straggler detection has come up quite a bit both internally and for use cases in OSS

**Comment by zaptrem** [comment]
Did some other PR meet this need? Trying to track down some stragglers but it seems discussion of them in general is rare across the PyTorch ecosystem.

**Comment by ezyang** [comment]
The way we do this internally is we have a high performance structured logging system (Scuba) and we send data from all ranks there and then can look for patterns across multiple ranks. Not sure what the best OSS equivalents here are though.

**Comment by ezyang** [comment]
cc @williamwen42 this is an example of OSS observability need

**Comment by abhinavsriva** [comment]
Hi all,

I started building an OSS prototype aligned with this RFC (TraceML: https://github.com/traceopt-ai/traceml/):

- Collects **per-rank, per-step** timings (forward / backward / optimizer)
- Focus is **cross-rank comparison**: identifying **skew / straggler ranks** across steps
- You can see **live run view**; the * summary report**  is still in progress

If you find interesting, please open issues with 
(a) common DDP is slow scenarios,
(b) what *minimal* signals would make this actionable (comm/wait vs data stalls vs compute).  

Right now it works on **single-node, multi-GPU DDP** and I am looking for feedback on what to improve next and which direction would be most useful.
